src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -625,4 +628,75 @@ case class TableFilter(child: TableIR, pred: IR) extends TableIR {
     val f = fb.result()
     ktv.filter((rv, globalRV) => f()(rv.region, rv.offset, false, globalRV.offset, false))
   }
+}
+
+case class TableAnnotate(child: TableIR, paths: IndexedSeq[List[String]], preds: IndexedSeq[IR]) extends TableIR {
```

Comment:
This begs the question what "level" the TableIR nodes should operate at:

1. should they operate at the user level as you've done, that is, the IR nodes roughly correspond to what the user writes for an expression,

2. at the opposite extreme, they could be expressions that consume and produce a singel value (the row), in which case the logic to build the nested stack of `InsertFields` would be outside the IR node and this IR node would essentially correspond to "map",

3. the intermediate solution that treats a table essentially as a struct, and TableAnnotation would essentially correspond to InsertFields, and instead of `paths: ISeq[List[String]]` you'd have `fields: ISeq[String]`.

I'm inclined against (1) since the user interface should be designed to maximize conciseness and expressiveness for the user while the IR should be designed to maximize the ability to be analyzed and optimized.  (3) has a nice symmetry between the struct operators in the IR and the table operators in the TableIR, but I think (2) will ultimately be simpler because we won't have to duplicate logic between the TableIR and IR.

The other nice (?)  thing about (2) is that it forces us to handle the analysis and optimization of structs at every level because the structs at the top level and deeply nested structs are expressed in the same way.  With (3), there's a temptation to give special treatment to the toplevel fields.

What do you think?